flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQDeserializationSchema.java

```diff
@@ -0,0 +1,78 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.rabbitmq;
+
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.typeutils.ResultTypeQueryable;
+import org.apache.flink.util.Collector;
+
+import com.rabbitmq.client.AMQP;
+import com.rabbitmq.client.Envelope;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.List;
+
+/**
+ * Interface for the set of methods required to parse an RMQ delivery.
+ * @param <T> The output type of the {@link RMQSource}
+ */
+public interface RMQDeserializationSchema<T> extends Serializable, ResultTypeQueryable<T> {
+	/**
+	 * This method takes all the RabbitMQ delivery information supplied by the client extract the data and pass it to the
+	 * collector.
+	 * NOTICE: The implementation of this method MUST call {@link RMQCollector#setMessageIdentifiers(String, long)} with
+	 * the correlation ID of the message if checkpointing and UseCorrelationID (in the RMQSource constructor) were enabled
+	 * the {@link RMQSource}.
+	 * @param envelope
+	 * @param properties
+	 * @param body
+	 * @throws IOException
+	 */
+	public  void processMessage(Envelope envelope, AMQP.BasicProperties properties, byte[] body, RMQCollector collector) throws IOException;
+
+	/**
+	 * Method to decide whether the element signals the end of the stream. If
+	 * true is returned the element won't be emitted.
+	 *
+	 * @param nextElement The element to test for the end-of-stream signal.
+	 * @return True, if the element signals end of stream, false otherwise.
+	 */
+	public boolean isEndOfStream(T nextElement);
+
+	/**
+	 * The {@link TypeInformation} for the deserialized T.
+	 * As an example the proper implementation of this method if T is a String is:
+	 * {@code return TypeExtractor.getForClass(String.class)}
+	 * @return TypeInformation
+	 */
+	public TypeInformation<T> getProducedType();
+
+
+	/**
+	 * Special collector for RMQ messages.
+	 * Captures the correlation ID and delivery tag also does the filtering logic for weather a message has been
+	 * processed or not.
+	 * @param <T>
+	 */
+	public interface RMQCollector<T> extends Collector<T> {
+		public void collect(List<T> records);
+
+		public void setMessageIdentifiers(String correlationId, long deliveryTag);
+	}
```

Comment:
What do you think about this interface instead?

```java

interface RMQCollector<T> extends Collector<T> {
		/**
		 * Called before a new message is processed in order to clear any temporary state.
		 */
		default void reset() {}

		void collect(String correlationId, long deliveryTag, T record);

		void collect(T record);
	}
```

Where instead of having the user call `setMessageIdentifiers`, we can call this ourselves from the implementation. We can also ensure the same `correlationId` and `deliveryTag` are used for a message by storing them in local state, and clear them when the next message comes with the `reset` method. This would follow the emitter pattern of other collectors a bit more and still allow users to call `collect` multiple times per message.

Reply:
+1, I was also thinking of an interface as the one described by @austince .

I am not sure though about the purpose of the `reset` method though. Especially with the default empy implementation. When would users call this method from `RMQCollector#deserialize` ? I think we might need that method in the implementation, but not necessarily in the interface.

Reply:
The point of having the method `setMessageIdentifiers` was the make sure that the user supplies the `correlationId` and `deliveryTag` before calling either `collect(T record)` or `collect(List<T> record` since those are used to validate that the record has not been added before.

If we go with the proposed interface we will have to call the `collect(String correlationId, long deliveryTag, T record)` method always and if you call `void collect(T record)` either run it without the checked because the `correlationId` and `deliveryTag` where not supplied or throw an error.

I also didn't understand the `reset()` method :D and it's relationship with the state. 

i'll keep the interface as is till you guys clarify this little bit of info for me.

Reply:
:) No worries on the `reset` method, I was just struggling to find a way to signal to the collector that it should clear the current correlation ID and delivery tag, if we're trying to ensure that processing a single message never calls `collect(String correlationId, long deliveryTag, T record)` with a different correlation ID and delivery tag. 

```java
// processMessage
AMQP.BasicProperties properties = delivery.getProperties();
byte[] body = delivery.getBody();
Envelope envelope = delivery.getEnvelope();
collector.reset(); // without this, how does the collector know a new message is being processed?
schema.deserialize(envelope, properties, body, collector);
```

For the question about the `collect(T record)`, what do you both think about throwing an exception if `autoAck` is false, since that means the user has configured checkpointing but their custom deserialization schema isn't handling the the `deliveryTag` and `correlationID`? The default schemas like `SimpleStringSchema` should still work, as they'll be wrapped to use `collect(String correlationId, long deliveryTag, T record)`.

Reply:
I think i finally understand your point. 
You're saying if there is a situation when 1 AMQP message received creates N messages to the collector 1:n there might be a situation where those N messages could have different corrlelation ids ? and there should be a way for them to be reset somehow ?

but then:
The suggestion above will not work since we create an instance of the collector every new message we get. So the reset is kinnda implied there.

Also you can achieve the 1:N with multiple correlation IDs but just calling the `setMessageIdentifiers` multiple times.

Let me know if got it right ... also that's my last working week then i have 2 weeks off .. 1 of them I'll be "safely" travelling :D the other one i intend to be creative and finish the 100 projects i am working on right now. One of the main topics is close this PR .. So I'll be like a car on the Autoban ... with No Speed Limits !!!! Vrooom Vroom

Reply:
@austince Generally I think you are right we might need the `clear` method. As I said before though, I think we should not expose that method in the interface of the Collector. It is an internal behaviour that users should not be concerned about. As @senegalo said theoretically we could even realize the `reset` method by creating a new instance of the collector for every incoming message. (For performance reasons we should not do it though, to mimimise the number of created objects).

My original thinking was that we can have different correlationId/deliveryTag for records deserialized from a single RMQ message. I am not an expert on RMQ though. If you are saying we should only ever have a single deliveryTag/correlationId for a single RMQ message I would go with an interface like:

```
interface RMQCollector<T> extends Collector<T> {
       // the two method can be called just once
    	void setCorrelationId(String correlationId);
        void setDeliveryTag(long deliveryTag);

        // can be called multiple times
		void collect(T record);
	}
```

In the implementation we would also have the `reset` method to do what @austince suggested:
```
AMQP.BasicProperties properties = delivery.getProperties();
byte[] body = delivery.getBody();
Envelope envelope = delivery.getEnvelope();
collector.reset(); // without this, how does the collector know a new message is being processed? Is it left up to the schema to reset the collector?
schema.deserialize(envelope, properties, body, collector);
```

Sorry that I am reverting back to one of the previous ideas, but now that I am thinking if we can only ever emit a single correlation id from a single RMQ message the `collect(String correlationId, int delveryTag, T record)` does not look right to me, as the user would have to keep the correlationId and deliveryTag around for subsequent records.


As for the exception when user does not provide the `deliveryTag` or/and `correaltionId`. What do you think about falling back to the current behaviour? We could simply automatically retrieve them from the properties/envelope. I think it will be the most common case.